refactor: move glErrorBar calculations from JavaScript into web-gl

jamesmporter opened the pull request:
Every web-gl attribute, defined in `glErrorBar.js`, now only varies with either `element` or `vertex` but not both.

Solves part of: https://github.com/d3fc/d3fc/issues/1371

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -16,35 +16,49 @@ export default () => {
+        (data, element) => data[element]
     );
 
+    // x-y coordinate to locate the "corners" of the element (ie errorbar). The `z` coordinate locates the corner relative to the line (this takes line width into account).
```
Multiline comments should use `/* */`

chrisprice replied:
Can you describe Z in the same way you've described the other two?

chrisprice replied:
Or is it not that simple?

jamesmporter replied:
It's a little more complex:

```
    /*
    x-y coordinate to locate the "corners" of the element (ie errorbar). The `z` coordinate locates the corner relative to the line (this takes line width into account).
    X: -1: LEFT, 0: MIDDLE, 1: RIGHT
    Y: 0: HIGH, 1: LOW
    
    if X==0 (corner part of a vertical line) {
      Z: 1: BOTTOM, -1: TOP
    } else (corner part of a horizontal line) {
      Z: 1=RIGHT, -1=LEFT
    }
    */
```

jamesmporter replied:
As per our discussion, I'll change Y to vary between -1 and 1 so that Z can follow X or Y.

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -16,35 +16,49 @@ export default () => {
-        .data([0, 0, 0, 0, 0, 0, 1, -1, -1, 1, 1, -1, -1, 1, 1, -1, -1, 1])
-        .value((data, element, vertex) => data[vertex]);
+
+    const yHighValueAttribute = projectedAttributeBuilder().value(
```
`highValue`

jamesmporter replied:
Do you mean `highValueAttribute`?

jamesmporter replied:
Done

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -16,35 +16,49 @@ export default () => {
+        (data, element) => data[element]
+    );
+
+    const yLowValueAttribute = projectedAttributeBuilder().value(
```
`lowValue`

jamesmporter replied:
Do you mean `lowValueAttribute`?

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -56,10 +70,10 @@ export default () => {
-            .attribute('aXDirection', xDirectionAttribute)
-            .attribute('aYDirection', yDirectionAttribute)
-            .attribute('aBandwidth', bandwidthAttribute);
+            .attribute('aYHighValueAttribute', yHighValueAttribute)
```
`aHighValue` (the `a` prefix identifies it as an attribute)

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -56,10 +70,10 @@ export default () => {
-            .attribute('aYDirection', yDirectionAttribute)
-            .attribute('aBandwidth', bandwidthAttribute);
+            .attribute('aYHighValueAttribute', yHighValueAttribute)
+            .attribute('aYLowValueAttribute', yLowValueAttribute)
```
`aLowValue`

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -56,10 +70,10 @@ export default () => {
+            .attribute('aYHighValueAttribute', yHighValueAttribute)
+            .attribute('aYLowValueAttribute', yLowValueAttribute)
+            .attribute('aBandwidth', bandwidthAttribute)
+            .attribute('aCornerAttribute', cornerCoordinateAttribute);
```
`aCorner`

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -16,35 +16,49 @@ export default () => {
+    // x-y coordinate to locate the "corners" of the element (ie errorbar). The `z` coordinate locates the corner relative to the line (this takes line width into account).
+    // X: -1: LEFT, 0: MIDDLE, 1: RIGHT
+    // Y: 0: HIGH, 1: LOW
+    const cornerCoordinateAttribute = projectedAttributeBuilder()
```
`cornerAttribute`

jamesmporter replied:
Done.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -16,35 +16,49 @@ export default () => {
+        ])
+        .value((data, element, vertex, component) => data[vertex][component]);
+
+    /**
```
Please remove the JSDoc style comments. We favour documenting public APIs in the `README.md` files for now.

chrisprice replied:
As this isn't yet finalised as an API, I'm fine with not documenting it for now.

jamesmporter replied:
Okay - removed.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -69,32 +83,48 @@ export default () => {
         program(numElements);
     };
 
+    /**
```
As above

jamesmporter replied:
Removed.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -69,32 +83,48 @@ export default () => {
         return draw;
     };
 
+    /**
```
As above

jamesmporter replied:
Removed.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -69,32 +83,48 @@ export default () => {
         return draw;
     };
 
+    /**
```
As above

jamesmporter replied:
Removed.

chrisprice commented on packages/d3fc-webgl/src/series/glErrorBar.js:
```diff
@@ -69,32 +83,48 @@ export default () => {
         return draw;
     };
 
+    /**
```
As above

jamesmporter replied:
Removed.

chrisprice commented on packages/d3fc-webgl/src/shaders/vertexShaderSnippets.js:
```diff
@@ -124,13 +124,25 @@ export const postScaleLine = {
         uniform float uLineWidth;`,
-    body: `gl_Position = vec4(aXValue, aYValue, 0, 1);`
+    body: `
+        // If "aCornerAttribute.y" is 0.0 then it represents a "high" corner, if 1.0 then it represents a "low" corner.
```
Please remove these comments as they duplicate the comment describing these values in `glErrorBar`. Duplication invariably leads to out-of-date comments which are worse than no comments!

jamesmporter replied:
Removed.

jamesmporter replied:
I've added `float isLow = aCorner.y` instead of this comment.

chrisprice reviewed (changes requested):
Implementation looks good, just stylistic comments. As a general rule I always favour consistency over *correctness*. 

I think there might also be some prettier failures, try `npm run eslint -- --fix`.

Commit:
refactor: move glErrorBar calculations from JavaScript into web-gl

Make each web-gl attribute vary with either the element index or vertex index but not both. This will enable a future optimisation where shorter arrays can be passed into web-gl.

jamesmporter commented:
@chrisprice I've addressed your comments and fixed the commit naming issue.

chrisprice commented on packages/d3fc-webgl/src/shaders/vertexShaderSnippets.js:
```diff
@@ -124,13 +124,25 @@ export const postScaleLine = {
+        float yValue = isLow * aLowValue + (1.0 - isLow) * aHighValue;
+
+        float isEdgeCorner = abs(aCorner.x);
+        float lineWidthXDirection = (1.0 - isEdgeCorner) * aCorner.z;     // If non-zero then this corner is part of a vertical line.
```
Isn't `z` always non-zero?

jamesmporter replied:
As discussed, I've removed the comments.
